Jolokia collector fixes

kasprzol opened the pull request:
We have encountered some problems with the Jolokia collector frequently timing out. The root cause was Jolokia's degraded performance when a connection wasn't closed (probably caused by signal in Diamond's scheduler interrupting the `urllib2.urlopen` call).

This pull request fixes this root cause (by adding timeouts to HTTP requests) as well as some of the slowdowns (e.g. caches regular expressions and the list of JMX domains).

bmhatfield commented:
cc @MichaelDoyle to take a look here.

@kasprzol - can you check out why CI is failing?

kasprzol commented:
@bmhatfield: too much rebasing, too litle sleep :) should be fixed now.

MichaelDoyle commented on src/collectors/jolokia/jolokia.py:
```diff
@@ -119,53 +163,75 @@ def check_mbean(self, mbean):
         return json.loads(json_str)
 
-    def list_request(self):
+    def _list_request(self):
```
If we're going to change this here, can we change the other "private" methods to use the "_" convention as well? We should just be consistent one way or another.

MichaelDoyle commented on src/collectors/jolokia/jolokia.py:
```diff
@@ -119,53 +163,75 @@ def check_mbean(self, mbean):
 
-    def list_request(self):
+    def _list_request(self):
+        """Retruns a dictionary with JMX domain names as keys"""
```
Spelling - should be "Returns" instead of "Retruns". :)

MichaelDoyle commented:
I think these all look like good changes. Thanks! Just made a couple small comments.

MichaelDoyle commented on src/collectors/jolokia/jolokia.py:
```diff
@@ -83,7 +87,16 @@ def get_default_config_help(self):
-            'path': 'Path to jolokia.  typically "jmx" or "jolokia"'
+            'path': 'Path to jolokia.  typically "jmx" or "jolokia"',
+            # https://github.com/rhuss/jolokia/blob/959424888a82abc2b1906c60547cd4df280f3b71/client/java/src/main/java/org/jolokia/client/request/J4pQueryParameter.java#L68
+            'use_cannonical_names': 'Whether property keys of ObjectNames'
```
Spelling `use_cannonical_names` should be `use_canonical_names`

MichaelDoyle commented on src/collectors/jolokia/jolokia.py:
```diff
@@ -98,20 +111,51 @@ def get_default_config(self):
             'password': None,
             'host': 'localhost',
             'port': 8778,
+            'use_cannonical_names': True,
```
Spelling `use_cannonical_names` should be `use_canonical_names`

MichaelDoyle commented on src/collectors/jolokia/jolokia.py:
```diff
@@ -72,6 +72,10 @@ def get_default_config_help(self):
         config_help = super(JolokiaCollector,
                             self).get_default_config_help()
         config_help.update({
+            'domains_to_check': "Pipe delimited list of JMX domains from which"
```
We can probably simply name this `domains`.

MichaelDoyle commented on src/collectors/jolokia/jolokia.py:
```diff
@@ -119,53 +163,75 @@ def check_mbean(self, mbean):
+            # restricts the return value to a map with the JMX domains as keys,
+            # a maxDepth of 2 truncates the map returned to the domain names
+            # (first level) and the MBean's properties (second level). The final
+            # values of the maps don't have any meaning and are dummy values.
```
We can probably simplify this comment to explain why we've chosen maxDepth=1 and refer to the doc for the rest of the details/options.

MichaelDoyle commented:
I think these all look like good changes. Thanks! Just made a couple small comments.

kasprzol commented:
Thanks for the feedback. I've fixed the issues you highlighted. I also added an option to separate the path component of metrics from the path component of the jolokia url.

kasprzol commented:
Hi

Are you planning on merging this in? Or do you have any other comments?

bmhatfield commented:
@MichaelDoyle let me know if you think this is all good to go, and I can merge it.

MichaelDoyle commented:
LGTM :key: :sparkles:
